### ISSUE TITLE
STOPWORDS in Topic Modelling

### ISSUE THREAD

**Issue by KMohaghegh** [chartbeat-labs/textacy#148]
Hi and thanks for your fantastic work :)
I am on unsupervised topic modelling with LDA and I have a problem. I want to extract stop words from the topics and I can do it in two ways:
1- remove the stop words from the raw text
2- ask the model to filter the stop words from the topics

none of those methods works for me.
I start with building the text corpus  for all of my text document one by one:

text = re.sub(u"@[A-Za-z0-9]+",u"",text)
text = textacy.preprocess.replace_urls(text, replace_with=u"")
text = textacy.preprocess.remove_punct(text, marks='-,!;:')
doc = textacy.Doc(unicode(text),lang=u'en', metadata=None)
corpus_new.add_doc(doc, metadata=None)
this works fine and build the corpus without url and punctuation. Could not find a way to get rid of the stop words here!

then I build up my model like this:

vectorizer = textacy.Vectorizer(weighting='tf', normalize=True, smooth_idf=True, min_df=0.07, max_df=0.9)
doc_term_matrix= vectorizer.fit_transform((doc.to_terms_list(ngrams=[1,2,3,4], filter_stops=True, named_entities=True, as_strings=True) for doc in corpus_new))
model = textacy.TopicModel('lda', n_topics=10)

a part of the  results is:

topic 2 : fund   small   not   social and governance   factor   4
topic 3 : market   have   fix   return   investor   invest

as you could see I filter out the STOPWORDS filter_stops=True but in the topic I have the word 'have' for example which is obviously a STOPWORDS.
Please help me where and how I should filter out the STOPWORDS.

Many Thanks
## Your Environment

* `textacy` and `spacy` versions: Both last versions 
* Python version: 2.7.13
* Operating system and version: MAC

**Comment by gryBox** [comment]
There has to be a better way but I do this:
```
stopWordsRepl_tupl = tuple((wrd, ' ') for wrd in stopWords_lst)
txt = reduce(lambda a, kv: a.replace(*kv), stopWordsRepl_tupl, txt)
```
put that code right before you convert the `text` to a `doc`

**Comment by KMohaghegh** [comment]
Thanks for your reply, it works but it removes also part of the word which are common with stop words like "hello" and then it removes "he" from it which is obvious a stop words but not inside "hello"! do you have any idea how to fix it.

**Comment by bdewilde** [comment]
Hi @KMohaghegh , it's hard to know exactly what'll work in any given situation, but here are some general recommendations:

- *Don't* remove punctuation as preprocessing of the raw text. spaCy uses that information to tag and parse the text, and your results downstream may be affected by it.
- *Do* remove stop words and punctuation when converting the parsed documents into lists of terms, i.e. `doc.to_terms_list(filter_stops=True, filter_punct=True)`.
- *Do* filter out the most and least common words by frequency in the vectorizer; specifying `max_df=0.9` as you do should remove all terms that occur in more than 90% of documents, which is a pretty aggressive guard against stopwords

I can check if there's a bug somewhere that's somehow letting stopwords get through filtering; pretty sure I have tests for this, though, so I'm doubtful. Will let you know if I find anything, but in the meantime, I'm going to close out this issue. Feel free to re-open if you think something is definitely wrong with the code.

**Comment by KMohaghegh** [comment]
Thanks for your comment and nice job :)
